pkg/plugins/manager.go

```diff
@@ -0,0 +1,113 @@
+package plugins
+
+import (
+	"errors"
+	"os"
+	"os/exec"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-plugin"
+	"github.com/treeverse/lakefs/pkg/logging"
+)
+
+var (
+	ErrPluginOfWrongType    = errors.New("plugin of the wrong type")
+	ErrPluginNameNotFound   = errors.New("unknown plugin name")
+	ErrUninitializedManager = errors.New("uninitialized plugins manager")
+)
+
+var allowedProtocols = []plugin.Protocol{
+	plugin.ProtocolGRPC,
+}
+
+// PluginIdentity identifies the plugin's version and executable location.
+type PluginIdentity struct {
+	Version int
+	// Cmd is the command that is used to run the plugin executable on the local filesystem.
+	Cmd *exec.Cmd
+}
+
+// PluginHandshake includes handshake properties for the plugin.
+type PluginHandshake struct {
+	Key   string
+	Value string
+}
+
+// Manager maps the available plugin names to the for different kinds of plugin.Client plugin controllers.
+// T is the custom interface type that the returned GRPCClient implementation implements, e.g. "Differ" for `plugin.Client`s that
+// include a GRPCClient that implements the "Differ" interface:
+// grpcPluginClient, err := c.Client() // Returns a plugin.GRPCClient
+// rawGrpcClientStub, err := grpcPluginClient.Dispense(name) // Calls grpcPluginClient's GRPCClient method and returns the gRPC stub.
+// grpcClient, ok := rawGrpcClient.(Differ) // Asserts the expected type of stub client.
+//
+// The map might include a mapping of "delta" -> plugin.Client to communicate with the Delta plugin.
+type Manager[T any] struct {
+	pluginApplicationClients map[string]*plugin.Client
+}
+
+func NewManager[T any]() *Manager[T] {
+	return &Manager[T]{
+		pluginApplicationClients: make(map[string]*plugin.Client),
+	}
+}
+
+// RegisterPlugin is used to register a new plugin client with the corresponding plugin type.
+func (m *Manager[T]) RegisterPlugin(name string, id PluginIdentity, auth PluginHandshake, p plugin.Plugin) error {
+	if m == nil || m.pluginApplicationClients == nil {
+		return ErrUninitializedManager
+	}
+	hc := plugin.HandshakeConfig{
+		ProtocolVersion:  uint(id.Version),
+		MagicCookieKey:   auth.Key,
+		MagicCookieValue: auth.Value,
+	}
+	cmd := id.Cmd
+	c := newPluginClient(name, p, hc, cmd)
+	m.pluginApplicationClients[name] = c
+	return nil
+}
+
+func newPluginClient(name string, p plugin.Plugin, hc plugin.HandshakeConfig, cmd *exec.Cmd) *plugin.Client {
+	clientConfig := plugin.ClientConfig{
+		Plugins: map[string]plugin.Plugin{
+			name: p,
+		},
+		AllowedProtocols: allowedProtocols,
+		HandshakeConfig:  hc,
+		Cmd:              cmd,
+	}
+	hl := hclog.New(&hclog.LoggerOptions{
+		Name:   name,
+		Output: os.Stdout,
+		Level:  hclog.Debug,
+	})
+	l := logging.Default()
+	hcl := NewHClogger(hl, l)
+	clientConfig.Logger = hcl
+	return plugin.NewClient(&clientConfig)
+}
+
+// LoadPluginClient loads a Client of type T.
+func (m *Manager[T]) LoadPluginClient(name string) (*T, error) {
```

Comment:
Consider spec'cing the manager on `*T` when you create it, and returning `T` here.  For instance, if `T` is an interface (it often will be!) then this is a pointer-to-interface, which is really weird.  Even if the _current_ implementation in the library returns a struct, there is no requirement for it to stay that way.

Reply:
What do you mean by `Consider **spec'cing** the manager on *T `

Reply:
The type `Manager[T]` returns clients of type `*T`.  This is technically correct if the client is a struct, and we instantiate it on the struct type:

```go
type MyClient struct { /* stuff */ } // But really the client is *MyClient

func a(m Manager[MyClient], name string) {
	var client *MyClient = m.LoadPluginClient(name)
	// client stuff
```

This works mostly because `*MyClient` satisfies the `MyClient` interface (the converse is not really true!).

BUT if `T` is an interface type, then `Manager[T].LoadPluginClient` returns a `*T`, which is a pretty useless type.

AFAIU the more Go-like interface would be to return a `T` from `Manager[T].LoadPluginClient`.  In the `*MyClient` case above, you would need to use `Manager[*MyClient]` as the manager type.  But that makes sense: the client type really is `*MyClient` there!

Reply:
got it